### DISCUSSION TITLE
[@types/aws-lambda] APIGatewayEventRequestContextJWTAuthorizer should guard against undefined `claims` key lookups

### DISCUSSION THREAD

**Discussion by dan-lovelace** [DefinitelyTyped/DefinitelyTyped#71043] [Issues with a @types package] [answered]
I'm using AWS Cognito to handle authn in my web server. The claims have groups attached to a key called `cognito:groups`.

I recently had a bug come up where a legacy user tried to sign in but was unable. The issue ended up being in my code around where I was assuming `cognito:groups` would always exist. The `APIGatewayEventRequestContextJWTAuthorizer` looks like this:

```ts
/**
 * JWT Authorizer Payload
 */
export interface APIGatewayEventRequestContextJWTAuthorizer {
    principalId: string;
    integrationLatency: number;
    jwt: {
        claims: { [name: string]: string | number | boolean | string[] };
        scopes: string[];
    };
}
```

Notice the `claims` keys are missing `undefined` in the union so Typescript was perfectly happy with my code:

```ts
const claimsString = claims[JWT_GROUPS_KEY].toString();
```

Which resulted in the error: `TypeError: Cannot read properties of undefined (reading 'toString')`

Would it be possible to add `undefined` to the list of possible claim key values?

[a comment by an automated account is not shown]

**Comment by simonbuchan** [comment, marked as answer]
In *general* aws-lambda types don't add the undefined to record types. (Arguably you should be using `--noUncheckedIndexAccess`, but that has it's own issues)

This would be a significantly breaking change, and it has a better solution by, for example, making the claims object a type parameter (with a default!) so you can explicitly declare what claims you expect.

**Comment by dan-lovelace** [comment]
Makes sense @simonbuchan. Thanks for the quick reply! Did not know about `noUncheckedIndexAccess`, that does exactly what I was looking for.
